crates/rust-analyzer/src/lsp/from_proto.rs

```diff
@@ -35,9 +35,20 @@ pub(crate) fn offset(
                 .ok_or_else(|| format_err!("Invalid wide col offset"))?
         }
     };
-    let text_size = line_index.index.offset(line_col).ok_or_else(|| {
-        format_err!("Invalid offset {line_col:?} (line index length: {:?})", line_index.index.len())
-    })?;
+    let (text_size, clamped_length) =
+        line_index.index.offset_clamped(line_col).ok_or_else(|| {
+            format_err!(
+                "Invalid offset {line_col:?} (line index length: {:?})",
+                line_index.index.len()
+            )
+        })?;
+    if let Some(clamped_length) = clamped_length {
+        tracing::error!(
```

Comment:
If LSP allows that, should we warn at it?

The answer probably lies in whether there are some client that deliberately sends longer line lengths. But if not, then maybe we don't want to merge this...

Reply:
> The answer probably lies in whether there are some client that deliberately sends longer line lengths

Indirectly yes. But the more direct answer is in the specification (unless we want to deliberately deviate from that which would be surprising).
Ideally I'd prefer following the spec today but also remove this feature from the spec later (assuming no client needs it).
Since it's allowed by the spec I agree we shouldn't error. But I think a warning would be appropriate; if nothing else it will help us find out which client wants to use this feature.

Assuming we want to warn, we can define our own LSP protocol extension that clients can opt into.
Say we call it `strict-mode`.
Then we can suppress the warning by default and only print it if strict mode is enabled.
LSP is sometimes a bit underspecified, meaning that servers and clients have to implement multiple different ways the protocol could be interpreted. We could use a strict mode as a means to eventually simplify the spec..

I'd also be fine with simply dropping the warning but I guess I have a slight preference for keeping it.

This crash is not necessarily an important issue but I believe I've hit it occasionally over a few years and didn't notice the crash because my client restarts `rust-analyzer` automatically on crash. I got better error propagation set up now.

Reply:
I'd say lets keep the log for now, and see if any client is being spammy with this or not